src/backend/gsroa.c

```diff
@@ -49,15 +54,66 @@ static void sliceStructs_Gather(SymInfo *sia, elem *e)
                 if (si >= 0 && sia[si].canSlice)
                 {
                     assert(si < globsym.top);
-                    if (tysize(e->Ety) != REGSIZE ||
-                        (e->Eoffset != 0 && e->Eoffset != REGSIZE))
+                    unsigned sz = tysize(e->Ety);
+                    if (sz == 2 * REGSIZE)
+                    {
+                        // Rewrite as OPpair later
+                        sia[si].usePair = true;
+
+                        /* OPpair cannot handle XMM registers, cdpair() and fixresult()
+                         */
+                        if (tyfloating(sia[si].ty0) || tyfloating(sia[si].ty1))
```

Comment:
Yes, it just wasn't obvious how the two branches relate to each other.
They are independent b/c different parts of the expression tree can access the same symbol, one as slice, one as fat value using OPpair.